### DISCUSSION TITLE
can not disable proxy_request_buffering causing upload problem

### DISCUSSION THREAD

**Discussion by darkclip** [ZoeyVid/NPMplus#736] [Q&A] [answered]
I have a backend app that needs to directly handle large file upload post requests.
As far as I know, the only parameter that matters is to turn off proxy_request_buffering, and that's what I've always been doing. 
However, it seems no matter where I set the param, npmplus always buffers the client request body. And that cause the large upload fails.
I've also tried lots of other configs seem no effect. (disable http2 http3 modsec etc.) And I also tested on a different backend (sftpgo), the same result.
It seems this issue only happens with npmplus. Any ideas?

**Comment by Zoey2936** [comment]
How do you know that it is still bufferd?

**Comment by darkclip** [comment]
For starters, because client_body_buffer_size is set to a relatively small size, a warning of `a client request body is buffered to a temporary file` keeps popping up.
Second, the backend couldn't receive the post request until the entire file is sent.

**Reply by Zoey2936** [reply]
are you sure that proxy_request_buffering is the correct value you need?

**Reply by darkclip** [reply]
Yeah, I'm pretty sure. I've deployed the backend service with nginx or npm for a long time. I'm fairly familiar with those configs that I always use.
BTW, can you confirm that proxy_request_buffering can not be disabled in npmplus at the moment?
@Zoey2936

**Reply by Zoey2936** [reply]
it should work, so no idea

**Reply by darkclip** [reply]
for anyone who has the same problem, the cause of it is due to http 2.0 and above support.
because from nginx to backend, the max proxy version is 1.1, when http/2 request coming in to nginx, nginx will aways buffer request regardless of proxy_request_buffering setting.

**Reply by Zoey2936** [reply]
do you use crowdsec?

**Reply by darkclip** [reply]
yes, but I don't think crowdsec is relevant. 
upon further investigation, for nginx it's not possible to use different protocols for the same port.
so once set listen for HTTP2 or QUIC for default port, it seems not possible to turn off proxy_request_buffering for any server.

**Reply by Zoey2936** [reply, marked as answer]
It is relevant, because if you disable it, it should works. At least for proxy_buffering, but it should also fix proxy_request_buffering

**Reply by darkclip** [reply]
tried your suggestion, no luck. -_-

**Reply by grigoroffbenj** [reply]
Same issue here ! 
I have to upload large file and it's still buffering
![Capture d'écran 2025-06-03 145721](https://github.com/user-attachments/assets/2aca7864-63c4-4dac-8a3d-e1578831618c)

**Reply by Zoey2936** [reply]
do you use crowdsec?

**Reply by grigoroffbenj** [reply]
yes and crowdsec with appsec

**Reply by darkclip** [reply]
After some digging in Nginx source code, I found out the problem is HTTP3.
proxy_request_buffering dose not work at all under HTTP3 in Nginx, causing all those uploading problems.

**Reply by Zoey2936** [reply]
> yes and crowdsec with appsec

does it work if you temporary disable crowdsec?

**Reply by grigoroffbenj** [reply]
In fact, if i put ENABLE=FALSE on docker path /data/crowdsec/crowdsec.conf. it's work it will not buffer anymore.

**Reply by Zoey2936** [reply]
Yes, it seems crowdsec requires buffering

**Reply by grigoroffbenj** [reply]
really sad news ..

**Reply by grigoroffbenj** [reply]
in addition, we need to upload big file ~400Mo and it's fail by 403 there is any modsec, crowdsec, appsec rules for that ?

**Reply by Zoey2936** [reply]
It is probably modsec, if you don't know how to configure modsec, just leave it disabled

**Reply by grigoroffbenj** [reply]
For this host modsec(so coreruleset too), brotli and HTTP/3-QUIC is disabled.
